_posts/2019-11-12-customizing-episerver-content-delivery-api.md

```diff
@@ -0,0 +1,927 @@
+```csharp
+    [RoutePrefix("api/v1/page")]
+    public class CustomContentController : ApiController
+    {
+        private readonly IUrlResolver _urlResolver;
+        private readonly IContentModelMapper _mapper;
+        private readonly CustomContentResultService _contentResultService;
+
+        public CustomContentController(IUrlResolver urlResolver, IContentModelMapper mapper, CustomContentResultService contentResultService)
+        {
+            _urlResolver = urlResolver;
+            _mapper = mapper;
+            _contentResultService = contentResultService;
+        }
+
+        [HttpGet]
+        [Route]
+        public IHttpActionResult Get([FromUri]string pageName)
+        {
+            // get the page content someway
+            UrlBuilder urlBuilder = new UrlBuilder($"http://localhost:54429/en/{pageName}");
+            IContent content = _urlResolver.Route(urlBuilder, ContextMode.Default);
+
+            // Transform IContent to ContentApiModel
+            ContentApiModel model = _mapper.TransformContent(content, false, "*");
+
+            // Filter out fields that are not needed
+            IDictionary<string, object> reducedModel = _contentResultService.ReduceFields(model);
+
+            // Add a custom field
+            reducedModel.Add("Breadcrumb", new []{ "Top", "Sub", "Bottom"});
+
+            return Ok(reducedModel);
+        }
+    }
+```
+It also gave me an ability to add a custom breadcrumb field to the final model.
+The drawback that you are now on your own, and need to worry yourself about request model validation, authorization, and some epi services that might not be invoked anymore.
+If you decide to go for your controller, do not forget to disable the original Episerver's content delivery endpoints.
+```csharp
+    [ModuleDependency(typeof(ContentApiCmsInitialization))]
+    public class ExtendedContentApiCmsInitialization : IConfigurableModule
+    {
+        public void Initialize(InitializationEngine context)
+        {
+        }
+        public void Uninitialize(InitializationEngine context)
+        {
+        }
+        public void ConfigureContainer(ServiceConfigurationContext context)
+        {
+            context.Services.Configure<ContentApiConfiguration>(config =>
+            {
+                config.Default()
+                    .SetSiteDefinitionApiEnabled(false)
+                    .SetMultiSiteFilteringEnabled(false);
+            });
+        }
+    }
+```
+## Sum 
+Now when I've written all this, the whole process looks very straight and easy. It was not that obvious when I've started with it. For a long period, I was considering dropping the use of Episerver's content delivery API classes and do everything myself. It might have even taken less time. Anyway, now I feel quite comfortable with the level of control that I have over the models. I hope this article might save some time for others.
```

Comment:
straight => straightforward
It was not that obvious when I've started => It was not that obvious when I started